opm/test_util/EclFilesComparator.cpp

```diff
@@ -305,3 +217,402 @@ double ECLFilesComparator::average(const std::vector<double>& vec) {
+            printValuesForCell(keyword, occurrence1, occurrence2, cell, data1, data2);
+            OPM_THROW(std::runtime_error, "Values of char type differ.");
+        }
+    }
+}
+
+
+
+void RegressionTest::intComparisonForOccurrence(const std::string& keyword, int occurrence1, int occurrence2) const {
+    ecl_kw_type* ecl_kw1 = nullptr;
+    ecl_kw_type* ecl_kw2 = nullptr;
+    const unsigned int numCells = getEclKeywordData(ecl_kw1, ecl_kw2, keyword, occurrence1, occurrence2);
+    std::vector<int> values1(numCells), values2(numCells);
+    ecl_kw_get_memcpy_int_data(ecl_kw1, values1.data());
+    ecl_kw_get_memcpy_int_data(ecl_kw2, values2.data());
+    for (size_t cell = 0; cell < values1.size(); cell++) {
+        if (values1[cell] != values2[cell]) {
+            printValuesForCell(keyword, occurrence1, occurrence2, cell, values1[cell], values2[cell]);
+            OPM_THROW(std::runtime_error, "Values of int type differ.");
+        }
+    }
+}
+
+
+
+void RegressionTest::doubleComparisonForOccurrence(const std::string& keyword, int occurrence1, int occurrence2) {
+    ecl_kw_type* ecl_kw1 = nullptr;
+    ecl_kw_type* ecl_kw2 = nullptr;
+    const unsigned int numCells = getEclKeywordData(ecl_kw1, ecl_kw2, keyword, occurrence1, occurrence2);
+    std::vector<double> values1(numCells), values2(numCells);
+    ecl_kw_get_data_as_double(ecl_kw1, values1.data());
+    ecl_kw_get_data_as_double(ecl_kw2, values2.data());
+
+    auto it = std::find(keywordDisallowNegatives.begin(), keywordDisallowNegatives.end(), keyword);
+    for (size_t cell = 0; cell < values1.size(); cell++) {
+        deviationsForCell(values1[cell], values2[cell], keyword, occurrence1, occurrence2, cell, it == keywordDisallowNegatives.end());
+    }
+}
+
+
+
+void RegressionTest::deviationsForCell(double val1, double val2, const std::string& keyword, int occurrence1, int occurrence2, size_t cell, bool allowNegativeValues) {
+    double absTolerance = getAbsTolerance();
+    double relTolerance = getRelTolerance();
+    if (!allowNegativeValues) {
+        if (val1 < 0) {
+            if (std::abs(val1) > absTolerance) {
+                printValuesForCell(keyword, occurrence1, occurrence2, cell, val1, val2);
+                OPM_THROW(std::runtime_error, "Negative value in first file, "
+                        << "which in absolute value exceeds the absolute tolerance of " << absTolerance << ".");
+            }
+            val1 = 0;
+        }
+        if (val2 < 0) {
+            if (std::abs(val2) > absTolerance) {
+                printValuesForCell(keyword, occurrence1, occurrence2, cell, val1, val2);
+                OPM_THROW(std::runtime_error, "Negative value in second file, "
+                        << "which in absolute value exceeds the absolute tolerance of " << absTolerance << ".");
+            }
+            val2 = 0;
+        }
+    }
+    Deviation dev = calculateDeviations(val1, val2);
+    if (dev.abs > absTolerance && dev.rel > relTolerance) {
```

Comment:
&& was the intention. the idea is that the absolute test 'should' fail in general, so the relative governs the general behavior. the exception is around 0 where the relative test will fail in general, the absolute test then 'runs the show'.